proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+
+Once a metadata provider was set up, when a log statement is about to be emitted, the logger will obtain the task-local `Baggage.current`,
+and pass it to the metadata providers. The providers can then convert specific baggage items into logger metadata as they see fit.
+
+Those metadata will then be included in the log statement, e.g. like this:
+
+```swift
+var baggage = Baggage.topLevel
+baggage.spanContext = SpanContext()
+Baggage.$current.withValue(baggage) {
+    test()
+}
+
+func test() {
+    log.info("Test", metadata: ["oneOff": "42"])
+    // info [traceID: abc, spanID: 123, onOff: 42] Test
+}
+```
+
+### Multiple `MetadataProvider`s using `MetadataProvider.multiplex(_:)`
+
+Borrowing the concept from log handlers, metadata providers also have a multiplexing implementation.
+It is defined as an extension on `MetadataProvider` and is useful in cases where users want to utilize
+more than one metadata provider at the same time:
+
+```swift
+extension Logger.MetadataProvider {
+    public static func multiplex(_ providers: [Logger.MetadataProvider]) -> Logger.MetadataProvider {
+        assert(!providers.isEmpty, "providers MUST NOT be empty")
+        return Logger.MetadataProvider { baggage in
+            providers.reduce(into: nil) { metadata, provider in
+                if let providedMetadata = provider.metadata(baggage) {
+                    if metadata != nil {
+                        metadata!.merge(providedMetadata, uniquingKeysWith: { _, rhs in rhs })
+                    } else {
+                        metadata = providedMetadata
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+Metadata keys are unique, so in case multiple metadata providers return the same key,
+the last provider in the array "wins" and provides the value.
+
+### Proposal: Using `Baggage` as the execution context dependent metadata container
+
+We propose that swift-log should use [Swift Distributed Tracing Baggage](https://github.com/apple/swift-distributed-tracing-baggage) for its contextual metadata propagation type, meaning that swift-log would depend on this package.
+
+The `Baggage` type is modeled after the [Propagation format for distributed context: Baggage](https://www.w3.org/TR/baggage/), a W3C working draft, and is intended for all kinds of instrumentation of Swift applications both within and across processes. It is also used by [Swift Distributed Tracing](https://github.com/apple/swift-distributed-tracing), however it is usable even without tracing infrastructure. The baggage package contains the single `Baggage` type, and is expected to have a stable, unchanging API, fit for swift-log to depend upon.
+
+> NOTE: We could potentially even rename the baggage package, if the tracing wording in it is causing confusion here. But realistically, this is its primary purpose: to enable tracing in logging systems, so we think swift-log depending on `swift-distributed-tracing-baggage` is fine.
```

Comment:
For me, the word "distributed" causes more confusion since logging doesn't necessarily have anything to do with "distributed".

Reply:
I agree. With this proposal the primary purpose of the package somehow changes since now it is not only used for tracing but also for providing log metadata through a task local. Can we imagine other use-cases where we want to use the baggage?

Just thinking out loud here what about `swift-task-baggage`? Though that might be a bit to abstract.

Reply:
Personally I'd be _very_ fine with that. To focus on a package that is a type good for "carrying around" whatever you want, and the existing infrastructure to propagate baggage -- regardless what is contained in it makes a lot of sense; callbacky frameworks picking it up and restoring it.

We had quite some naming drama bout the package last time, but I very much would like to make sure we all like what we end up with -- so that's on the table. Another idea was by @tomerd to sink it into swift-log, which would work, but I'm a bit worried about marrying it to logging _only_... Let's weigh our options here -- certianly the name you propose @FranzBusch would be quite good!